### PR TITLE
Header unions

### PR DESCRIPTION
**Pull request by jnfoster**
This is based on the text from #139.

### CODE REVIEW

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 1]
```diff
@@ -2231,6 +2253,181 @@ manipulate the elements at the front and back of the stack:
...
+header H2 {
+  bit<16> g;
+}  
+
+header_union U {
+  H1 h1;
+  H2 h2;
+}
+
+U u; // u invalid
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: Done.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 2]
```diff
@@ -2231,6 +2253,181 @@ manipulate the elements at the front and back of the stack:
   ```count``` argument must be a positive integer that is a
   compile-time known value. The return type is ```void```.
 
+## Operations on Header Unions { #sec-expr-hu }
+
+A variable declared with a union type is initially invalid:
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: Done.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 3]
```diff
@@ -2231,6 +2253,182 @@ manipulate the elements at the front and back of the stack:
...
+parser top(packet_in b, out Parsed_packet p) {
+    state start {
+       b.extract(p.ethernet);
+       transition select(p.ethernet.etherType) {
+           16w0x0800 : parse_ipv4;
+           16w0x86DD : parse_ipv6;
+       }
+   }
+   state parse_ipv4 {
+       b.extract(p.ip.ipv4);
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: Removed the check, per the 4/24 discussion and added an issue on p4c.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 4]
```diff
@@ -2231,6 +2253,181 @@ manipulate the elements at the front and back of the stack:
...
+- `u.hi.setValid():` sets the valid bit for header `hi` to true and sets
+  the valid bit for all other headers to false (which implies that
+  reading these headers will return an unspecified value).
+
+- `u.hi.setInvalid()`: if the valid bit for `hi` is true then sets it to
+  false (which implies that reading u.hi will return an unspecified
+  value); if the valid bit for `hi` is already false, then this
+  expression has no effect. In particular, if `u.hj` was valid for some `hj` other than `hi` before `u.hi.setInvalid()` was invoked, then `u.hj` remains valid and its fields continue to have the same values as before.
+
+We can understand an assignment to a union
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: Done.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 5]
```diff
@@ -1427,6 +1428,28 @@ Mpls_h[10] mpls;
 ~ End P4Example
 introduce a header stack called ```mpls``` containing 10 entries, each of type ```Mpls_h```.
 
+### Header Unions {#sec-header-unions}
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: Done.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 6]
```diff
@@ -1427,6 +1428,28 @@ Mpls_h[10] mpls;
...
+
+A header union is defined as:
+~ Begin P4Grammar
+headerUnionDeclaration
+    : optAnnotations HEADER_UNION name
+      '{' structFieldList '}'         
+    ;
+~ End P4Grammar
+
+This declaration introduces a new type with the specified name in the local scope. Each element of the list of fields used to declare a header union must be a header type. However, the empty list of fields is legal.
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: We agreed to have empty unions in the 4/24 LDWG.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 7]
```diff
@@ -2246,6 +2269,181 @@ manipulate the elements at the front and back of the stack:
...
+
+This also implies that each of the headers `h1` through `hn` contained
+in it is also invalid. Unlike headers, a union cannot be initialized.
+
+However, we can change the validity of a header union by assigning a
+valid header to one of its elements:
+~Begin P4Example
+header H1
+U u; 
+H1 my_h1 = { 8w0 }; // my_h1 is valid
```
Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: I support having a similar example with tuples assigned to headers there. But let's keep this here.

File: p4-16/spec/P4-16-draft-spec.mdk (hunk 5 above)
Comment (jafingerhut) [inline p4-16/spec/P4-16-draft-spec.mdk]: There is another table that @mbudiu-vmw may have been thinking of, that it seems you may not have modified @jnfoster.  I have a separate PR for that one now: https://github.com/p4lang/p4-spec/pull/190

Comment (jnfoster) [inline p4-16/spec/P4-16-draft-spec.mdk]: I think I was git-challenged today. I swear I added it but I don't see it here now or in my local repository. In any event, #190 looks good. Let's merge them both once somebody signs off on this one.

**Pull request merged**